### PR TITLE
Abstractive Summarization using Deep Learning

### LINKED ISSUE #3542: Abstractive Summarization using Deep Learning
### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article higher ranking.*

Abstractive Summarization using Deep Learning

### Introduction paragraph (2-3 paragraphs):

Text summarization aims at generating accurate and concise summaries from input documents. In contrast, abstractive summarization may generate novel words. A good abstractive summary covers principal information in the input and is linguistically fluent.

The [Pegasus](https://arxiv.org/pdf/1912.08777.pdf) model is built using a Transformer Encoder-Decoder architecture and is ridiculously powerful when it comes to summarizing big blocks of text. In Pegasus, important sentences are removed/masked from an input document and are generated together as one output sequence from the remaining sentences. This tutorial will walk you through how use the Pegasus model to perform abstract summarization from start to finish. We will perform abstractive summarization on some Wikipedia, News and Scientific Journals documents.

### Key takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article? Use this area to get your ideas down on the bulk of your article or tutorial.

By the end of this article, a reader should be able to:

1. Install Dependencies for Transformers in Python
2. Import and Configure the Pegasus Model
3. Perform abstractive summarization on Wikipedia, News and Scientific Journals

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

- [PEGASUS: Pre-training with Extracted Gap-sentences for Abstractive Summarization](https://arxiv.org/pdf/1912.08777.pdf)

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by smkinyan**
This closes #3542

### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #issue-number`

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by hectorkambow** [comment]
[Abstractive Summarization Using Deep Learning #3651.pdf](https://github.com/section-engineering-education/engineering-education/files/7117881/Abstractive.Summarization.Using.Deep.Learning.3651.pdf)
 plagiarism check done - it will benefit the article to review the red flags in the repot.

**Review by collins-a — CHANGES_REQUESTED** [review]
Hey @smkinyan, simple tutorial on Abstractive Summarization. However, there are a few changes required as highlighted by my comments.

Ping me when done.

File: content/articles/abstractive-summarization-using-deep-learning/index.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+Let's view our tokens. We achieve this by writing the following command:
+
+```python
+tokens
+```
+Let's now go ahead and try to summarize this text.
+
+```python
+summary = model.generate(**tokens)
```
[the remark this answers is not in the document]
Comment (smkinyan) [reply content/articles/abstractive-summarization-using-deep-learning/index.md:139]: Solved

File: content/articles/abstractive-summarization-using-deep-learning/index.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+This command successfully installs PyTorch.
+
+Let's install our second dependency, which is the HuggingFace transformers.
+
+```python
+!pip3 install transformers
+```
+
+This command will successfully install the huggingface transformer library.
```
Comment (collins-a) [inline content/articles/abstractive-summarization-using-deep-learning/index.md:67]: Consider removing this sentence

File: content/articles/abstractive-summarization-using-deep-learning/index.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+tokens = tokenizer_model(text, truncation=True, padding="longest", return_tensors="pt")
+```
+We've created a variable called `tokens` to store our token representation. We use our tokenizer `tokenizer_model`, which we created earlier, to tokenize our texts. The `truncation=True` argument allows the model to truncate our texts into a size suitable for input into our model. The `return_tensors` argument tells the model to use tensors from PyTorch.
+
+Let's view our tokens. We achieve this by writing the following command:
+
+```python
+tokens
+```
+Let's go ahead and try to summarize this text.
```
Comment (collins-a) [inline content/articles/abstractive-summarization-using-deep-learning/index.md:135]:
```suggestion
It's time to summarize our text.
```

File: content/articles/abstractive-summarization-using-deep-learning/index.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+```python
+summary = loaded_model.generate(**tokens)
+```
+The `**tokens` unpacks our tokens and pass them into our model. The asterisks in `**tokens` are simply adding the `input_ids` and `attention_mask` present in the results above. 
+
+Let's check our generated summary results. We do this by typing the following command:
```
Comment (collins-a) [inline content/articles/abstractive-summarization-using-deep-learning/index.md:142]:
```suggestion
To check our generated summary results in tokens, we type `summary`.
```

File: content/articles/abstractive-summarization-using-deep-learning/index.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+```python
+summary
+```
+These results represent our output tensors. This is our summary in tokens.
+
+These may be just a bunch of numbers to humans, but not to machines. This is how they understand language. Decoding these values will help us make sense of these numbers.
```
Comment (collins-a) [inline content/articles/abstractive-summarization-using-deep-learning/index.md:149]:
```suggestion
These results represent our output tensors. These may be just a bunch of numbers to humans, but not to machines. This is how they understand language. Decoding these values will help us make sense of these numbers.
```

File: content/articles/abstractive-summarization-using-deep-learning/index.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+```python
+tokenizer_model.decode(summary[0])
+```
+The `summary` result above shows that the results are in a nested list. But we only need the first result, so that's why we've indexed [0].
+
+Running that command shows our summarized text. 
+
+Impressive, right?
+
+All that big block of text summarized into that short text.
```
Comment (collins-a) [inline content/articles/abstractive-summarization-using-deep-learning/index.md:160]:
```suggestion
Running the above command shows our summarized text. Isn't it impressive how that block of text has been summarized?
```

File: content/articles/abstractive-summarization-using-deep-learning/index.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+```
+The `summary` result above shows that the results are in a nested list. But we only need the first result, so that's why we've indexed [0].
+
+Running that command shows our summarized text. 
+
+Impressive, right?
+
+All that big block of text summarized into that short text. 
+
+So, here's the actual test. If you go through the huge block of text, you'll find that this summarized version of the text doesn't exist. It's novel and has been completely generated by the model.
```
Comment (collins-a) [inline content/articles/abstractive-summarization-using-deep-learning/index.md:162]: The beginning of this sentence is a bit unclear. A reader will likely ask, what exactly is being tested?

**Review by collins-a — APPROVED** [review]

**Comment by LinusMuema** [comment]
Hi @smkinyan 
Thank you for this amazing article. We'd like to see more of such great content.

cc: @collins-a
